contracts/pools/rai/StableSwapRAI.vy

```diff
@@ -0,0 +1,1136 @@
+                assert convert(response, bool)  # dev: failed transfer
+            # end "safeTransferFrom"
+
+    # Mint pool tokens
+    CurveToken(lp_token).mint(msg.sender, mint_amount)
+
+    log AddLiquidity(msg.sender, _amounts, fees, D1, token_supply + mint_amount)
+
+    return mint_amount
+
+
+@view
+@internal
+def _get_y(i: int128, j: int128, x: uint256, _xp: uint256[N_COINS]) -> uint256:
+    """
+    Calculate x[j] if one makes x[i] = x
+
+    Done by solving quadratic equation iteratively.
+    x_1**2 + x_1 * (sum' - (A*n**n - 1) * D / (A * n**n)) = D ** (n + 1) / (n ** (2 * n) * prod' * A)
+    x_1**2 + b*x_1 = c
+
+    x_1 = (x_1**2 + c) / (2*x_1 + b)
+    """
+    # x in the input is converted to the same price/precision
+
+    assert i != j       # dev: same coin
+    assert j >= 0       # dev: j below zero
+    assert j < N_COINS  # dev: j above N_COINS
+
+    # should be unreachable, but good for safety
+    assert i >= 0
+    assert i < N_COINS
+
+    A: uint256 = self._A()
+    D: uint256 = self._get_D(_xp, A)
+    Ann: uint256 = A * N_COINS
+    c: uint256 = D
+    S: uint256 = 0
+    _x: uint256 = 0
+    y_prev: uint256 = 0
+
+    for _i in range(N_COINS):
+        if _i == i:
+            _x = x
+        elif _i != j:
+            _x = _xp[_i]
+        else:
+            continue
+        S += _x
+        c = c * D / (_x * N_COINS)
+    c = c * D * A_PRECISION / (Ann * N_COINS)
+    b: uint256 = S + D * A_PRECISION / Ann  # - D
+    y: uint256 = D
+    for _i in range(255):
+        y_prev = y
+        y = (y*y + c) / (2 * y + b - D)
+        # Equality with the precision of 1
+        if y > y_prev:
+            if y - y_prev <= 1:
+                return y
+        else:
+            if y_prev - y <= 1:
+                return y
+    raise
+
+
+@view
+@external
+def get_dy(i: int128, j: int128, _dx: uint256) -> uint256:
+    rates: uint256[N_COINS] = [self._get_scaled_redemption_price(), self._vp_rate_ro()]
+    xp: uint256[N_COINS] = self._xp(rates[MAX_COIN])
+
+    x: uint256 = xp[i] + (_dx * rates[i] / PRECISION)
+    y: uint256 = self._get_y(i, j, x, xp)
+    dy: uint256 = xp[j] - y - 1
+    fee: uint256 = self.fee * dy / FEE_DENOMINATOR
+    return (dy - fee) * PRECISION / rates[j]
+
+
+@view
+@external
+def get_dy_underlying(i: int128, j: int128, _dx: uint256) -> uint256:
+    # dx and dy in underlying units
+    vp_rate: uint256 = self._vp_rate_ro()
+    xp: uint256[N_COINS] = self._xp(vp_rate)
+    base_pool: address = self.base_pool
+
+    # Use base_i or base_j if they are >= 0
+    base_i: int128 = i - MAX_COIN
+    base_j: int128 = j - MAX_COIN
+    meta_i: int128 = MAX_COIN
+    meta_j: int128 = MAX_COIN
+    if base_i < 0:
+        meta_i = i
+    if base_j < 0:
+        meta_j = j
+
+    x: uint256 = 0
+    if base_i < 0:
+        if i == REDEMPTION_COIN:
+            x = xp[i] + (_dx * self._get_scaled_redemption_price() / PRECISION)
+        else:
+            x = xp[i] + _dx
```

Comment:
I guess this line is never reached if we have 2-coin pools right? So it is just in case ever a 3-coin pool is needed?